src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs

```diff
@@ -0,0 +1,180 @@
+                }
+                lookupResult.Clear();
+                args.Add((member, expr));
+            }
+
+            if (receiverType is null || receiverType.IsVoidType())
+            {
+                diagnostics.Add(ErrorCode.ERR_InvalidWithReceiverType, syntax.Receiver.Location);
+                receiverType = CreateErrorType();
+            }
+
+            if (receiverType.IsErrorType())
+            {
+                lookupResult.Free();
+                return new BoundWithExpression(
+                    syntax,
+                    receiver,
+                    cloneMethod: null,
+                    args.ToImmutableAndFree(),
+                    receiverType,
+                    hasErrors: true);
+            }
+
+            // PROTOTYPE: The receiver type must have a instance method called 'Clone' with no parameters
+            LookupMembersWithoutInheritance(
+                lookupResult,
+                receiverType,
+                "Clone",
+                arity: 0,
+                LookupOptions.MustBeInstance | LookupOptions.MustBeInvocableIfMember,
+                this,
+                this.ContainingType,
+                diagnose: false,
+                ref useSiteDiagnostics,
+                ConsList<TypeSymbol>.Empty);
+
+            MethodSymbol? cloneMethod = null;
+            if (lookupResult.IsMultiViable)
+            {
+                foreach (var symbol in lookupResult.Symbols)
+                {
+                    if (symbol is MethodSymbol { ParameterCount: 0 } m)
+                    {
+                        cloneMethod = m;
+                        break;
+                    }
+                }
+            }
+
+            lookupResult.Clear();
+            useSiteDiagnostics = null;
+
+            if (cloneMethod is null)
+            {
+                hasErrors = true;
+                diagnostics.Add(ErrorCode.ERR_NoSingleCloneMethod, syntax.Receiver.Location, receiverType);
+            }
+            else
+            {
+                // Check return type
+                if (!receiverType.IsEqualToOrDerivedFrom(
+                        cloneMethod.ReturnType,
+                        TypeCompareKind.ConsiderEverything,
+                        ref useSiteDiagnostics))
+                {
+                    hasErrors = true;
+                    diagnostics.Add(
+                        ErrorCode.ERR_ContainingTypeMustDeriveFromWithReturnType,
+                        syntax.Receiver.Location,
+                        receiverType,
+                        cloneMethod.ReturnType);
+                }
+
+                useSiteDiagnostics = null;
+            }
+
+            lookupResult.Free();
+
+            return new BoundWithExpression(
+                syntax,
+                receiver,
+                cloneMethod,
+                args.ToImmutableAndFree(),
+                cloneMethod?.ReturnType ?? receiverType,
+                hasErrors: hasErrors);
```

Comment:
I cannot tell if this is correct as the specification does not tell me what the type of a with expression is.